packages/core/src/lib/utils/env.ts

```diff
@@ -51,13 +51,22 @@ export function createActionURL(
   envObject: any,
   basePath?: string
 ): URL {
-  let url = envObject.AUTH_URL ?? envObject.NEXTAUTH_URL
-  if (!url) {
-    const host = headers.get("x-forwarded-host") ?? headers.get("host")
-    if (!host) throw new TypeError("Missing host")
-    const proto = headers.get("x-forwarded-proto") ?? protocol
-    url = `${proto === "http" ? "http" : "https"}://${host}${basePath}`
+  let envUrl = envObject.AUTH_URL ?? envObject.NEXTAUTH_URL
+
+  let detectedHost = headers.get("x-forwarded-host") ?? headers.get("host")
+  let detectedProtocol = headers.get("x-forwarded-proto") ?? protocol ?? "https"
+
+  let url: URL
+  if (envUrl) {
+    url = new URL(envUrl)
+  } else {
+    url = new URL(`${detectedProtocol}://${detectedHost}`)
   }
 
-  return new URL(`${url.replace(/\/$/, "")}/${action}`)
+  // remove trailing slash
+  const sanitizedUrl = url.toString().replace(/\/$/, "")
+  // remove leading and trailing slash
+  const sanitizedBasePath = basePath?.replace(/(^\/|\/$)/g, "") ?? ""
+
+  return new URL(`${sanitizedUrl}/${sanitizedBasePath}/${action}`)
```

Comment:
I think we're missing the case where `url.pathname` equals to `basePath`. If so, we don't need to append the `basePath`. That should also a valid case IMO

Reply:
Based on earlier discussions ( https://github.com/nextauthjs/next-auth/pull/10094#issuecomment-1977096182 ) basepath shouldn't be part of the env url.

Seems unnecessary to support that since then you have to keep basepath in sync with envurl

Reply:
Technically, `basePath` is required in the env URL, but we supported it with or without before. You can see this explanation from @balazsorban44: https://github.com/nextauthjs/next-auth/issues/10081#issuecomment-1960626967

So in my opinion:
- If we detect both `AUTH_URL` and `basePath`, check if `AUTH_URL`'s `pathname` and `basePath` match:
  - If yes, we should not use `url.toString()`, but instead `const sanitizedUrl = url.origin.toString().replace(/\/$/, "")` we should also log a warning - my suggestion: `AUTH_URL and basePath detected. This is likely a configuration mistake - either remove the basePath, or remove the pathname of AUTH_URL.`.
  - Otherwise, construct the URL with `url.origin.toString()`, and log a warning: "AUTH_URL and basePath detected and are mismatching. This is likely a configuration mistake - either remove the basePath, or remove the pathname of AUTH_URL.\n basePath will override AUTH_URL pathname"
Basically, these two options should not be specified together, as they will overlap/ become redundant. 

If only basePath detected, use detectedHost + basePath
If only AUTH_URL detected, use use `url.toString()`

Hope this makes sense, my goal is to keep the existing setup from breaking at least at possible. I also discussed with @ndom91 before commenting 🙏

Reply:
@ThangHuuVu I believe it now supports the types of configurations expected.